plat/linuxu: Fix memory region setup and issue #766

mogasergiu opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [x] Updated relevant documentation.


### Base target

 - Architecture(s): [e.g. `x86_64` or N/A]
 - Platform(s): [e.g. `kvm`, `xen` or N/A]
 - Application(s): [e.g. `app-python3` or N/A]


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

Re-submission of #797. Therefore, #797 is superseeded by this.

RaduNichita reviewed (changes requested):
Hi, @mogasergiu!

This looks fine, I tested it with most of the complex apps and seems to be working OK. Thanks for continuing the previous work! :100:  

 Please also add the `Co-authored-by`: <> tag in the commit description, you can see the discussion from #880 and [here](https://github.com/unikraft/docs/pull/242)

RaduNichita reviewed (commented):
All good, now! Thanks, @mogasergiu :t-rex: 

Reviewed-by: Radu Nichita <radunichita99@gmail.com>

Daparrag committed:
plat/linuxu: Add check for memory region index out of range

Ensure that we do not try to fetch a memory region descriptor by an
index that is larger than our current memory region descriptor count.

Furthermore, this guarantees that `ukplat_memregion_count` is always
called before `ukplat_memregion_get` is executed fully. This way we
will always have the memory region descriptors as well as their count
set up and available before we try to fetch one.

Co-authored-by: Sergiu Moga <sergiu.moga@protonmail.com>
Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>
Signed-off-by: Diego Alejandro Parra Guzman <diego-alejandro.parra-guzman@tttech.com>
GitHub-Fixes: #766

Daparrag committed:
plat/linuxu: Avoid re-initialization of the `heap` and `initrd`

Keep a locally declared static variable `init` to cache the
current amount of memory region descriptors. This way, whenever
`ukplat_memregion_count` is called, `__linuxu_plat_heap_init` and
`__linuxu_plat_initrd_init` do not have to be executed more than
once.

Co-authored-by: Sergiu Moga <sergiu.moga@protonmail.com>
Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>
Signed-off-by: Diego Alejandro Parra Guzman <diego-alejandro.parra-guzman@tttech.com>

razvand reviewed (approved):
Thanks, @Daparrag, @mogasergiu.

Reviewed-by: Razvan Deaconescu <razvand@unikraft.io>
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
